src/physical/tabular/operations/triescan_prune.rs

```diff
@@ -0,0 +1,473 @@
+#[derive(Debug)]
+pub struct TrieScanPruneState<'a> {
+    /// Trie scan which is being pruned
+    pub input_trie_scan: TrieScanEnum<'a>,
+    initialized: bool,
+    /// Current column scan layer of the input trie scan
+    /// Layer zero is at to top of the trie scan
+    input_trie_scan_current_layer: usize,
+    /// This vector records for every column whether if has already
+    /// been peeked into by the `advance` function.
+    /// If so, the column should ignore the first call to `next` and
+    /// instead return the result of `current`.
+    /// This is required to implement the lookahead to check if
+    /// a value would exist in a materialized version of the trie scan.
+    column_peeks: Vec<bool>,
+    /// Layer on which an outside consumer would believe this trie scan to be
+    /// TODO: Maybe move to the `TrieScanPrune` itself
+    external_current_layer: usize,
+}
+
+impl<'a> TrieScanPruneState<'a> {
+    /// Decrements the [`external_current_layer`]
+    pub fn external_up(&mut self) {
+        debug_assert!(self.initialized);
+        assert!(self.external_current_layer > 0);
+
+        self.external_current_layer -= 1;
+    }
+
+    /// Increments the [`external_current_layer`]
+    pub fn external_down(&mut self) {
+        if !self.initialized {
+            self.initialized = true;
+            self.input_trie_scan.down();
+            return;
+        }
+
+        assert!(self.external_current_layer < self.input_trie_scan.get_types().len() - 1);
+
+        self.external_current_layer += 1;
+    }
+
+    fn up(&mut self) {
+        assert!(self.input_trie_scan_current_layer > 0);
+
+        self.input_trie_scan.up();
+        self.input_trie_scan_current_layer -= 1;
+    }
+
+    fn down(&mut self) {
+        assert!(self.input_trie_scan_current_layer < self.input_trie_scan.get_types().len() - 1);
+
+        self.input_trie_scan.down();
+        self.input_trie_scan_current_layer += 1;
+    }
+
+    /// Moves the input trie to the target layer through `up` and `down`
+    fn go_to_layer(&mut self, target_layer: usize) {
+        assert!(target_layer < self.input_trie_scan.get_types().len());
+
+        // Move up or down to the correct layer
+        // Note that the uppermost layer has index 0
+
+        for _ in self.input_trie_scan_current_layer..target_layer {
+            self.down();
+        }
+        for _ in target_layer..self.input_trie_scan_current_layer {
+            self.up();
+        }
+    }
+
+    /// Checks whether a column has been peeked already.
+    ///
+    /// See [`TrieScanPruneState`] for more information.
+    #[inline]
+    pub fn is_column_peeked(&self, index: usize) -> bool {
+        self.column_peeks[index]
+    }
+
+    #[inline]
+    unsafe fn get_input_trie_current_value(&mut self, index: usize) -> Option<StorageValueT> {
+        let scan = self.input_trie_scan.get_scan(index).unwrap();
+
+        unsafe { &mut *scan.get() }.current()
```

Comment:
Again, since you get `&mut self`, couldn't you use `get_mut` on the `UnsafeCell` without the `unsafe` block?
(Same comment for the method below.)